README.md
# BACnet Server MSTP Example C++

A minimal BACnet MSTP server example written in C++ using the [CAS BACnet Stack](https://store.chipkin.com/services/stacks/bacnet-stack). For a full featured BACnet Server example see the [BACnet Server Example CPP](https://github.com/chipkin/BACnetServerExampleCPP) project.

## Releases

Build versions of this example can be downloaded from the [Releases](https://github.com/chipkin/BACnetServerMSTPExampleCPP/releases) page.

## Installation

Download the latest release zip file on the [Releases](https://github.com/chipkin/BACnetServerMSTPExampleCPP/releases) page.

## Usage

The serial port, baud rate, and MAC address of this BACnet MSTP server can be configured using command line parameters:

```bash
BACnetServerMSTPExampleCPP [serial port] [baud rate] [mac address]
BACnetServerMSTPExampleCPP COM5 9600 1
BACnetServerMSTPExampleCPP ttyS6 19200 25
```

Device tree:

- Device: 389999 (Device Rainbow)
  - analog_value: 2 (AnalogValue Diamond)

## Build

A [Visual Studio 2019](https://visualstudio.microsoft.com/downloads/) project is included with this project. This project also auto built using [Gitlab CI](https://docs.gitlab.com/ee/ci/) on every commit.

1. Copy *CASBACnetStack_x64_Debug.dll*, *CASBACnetStack_x64_Debug.lib*, *CASBACnetStack_x64_Release.dll*, and *CASBACnetStack_x64_Release.lib* from the [CAS BACnet Stack](https://store.chipkin.com/services/stacks/bacnet-stack) project into the /bin/ folder.
2. Use [Visual Studio 2019](https://visualstudio.microsoft.com/vs/) to build the project. The solution can be found in the */BACnetServerMSTPExample/* folder.

A makefile that uses gcc is included for making executables on linux.

## Example Output

The application outputs all BYTES sent or received. BYTES with underscore after them are received bytes. BYTES with a space after them are sent bytes. When the user presses the "enter" key, the MSTP stats, states, and events are printed for reference.

```txt
CAS BACnet Stack Server MSTP Example v0.0.1.0
https://github.com/chipkin/BACnetServerMSTPExampleCPP

FYI: Using comport=[COM6], Portnr=[5]
FYI: Loading CAS BACnet Stack functions... OK
FYI: CAS BACnet Stack version: 3.12.11.0
FYI: Connecting serial resource to serial port. baudrate=[9600], ... OK
OK, Connected to port
FYI: Registering the callback Functions with the CAS BACnet Stack
Setting up server device. device.instance=[389999]
Created Device.
Enabling WriteProperty service... OK
Enabling device optional properties. Max_info_frames, max_master... OK
FYI: Sending I-AM broadcast
FYI: Entering main loop...
55_ff_01_03_0a_00_00_1b_  {src=0a, dest=03 POLL_FOR_MASTER}
55_ff_01_04_0a_00_00_a1_  {src=0a, dest=04 POLL_FOR_MASTER}
55_ff_01_05_0a_00_00_28_  {src=0a, dest=05 POLL_FOR_MASTER}
55_ff_01_06_0a_00_00_b0_  {src=0a, dest=06 POLL_FOR_MASTER}
[ReceivedPFM]
55 ff 02 0a 00 00 00 a0   {src=00, dest=0a REPLY_TO_POLL_FOR_MASTER}
55_ff_00_00_0a_00_00_05_  {src=0a, dest=00 TOKEN}
[SendCustomFrame]
55 ff 06 ff 00 00 15 16 01 20 ff ff 00 ff 10 00 c4 02 05 f3 6f 22 05 c4 91 03 22 01 85 ec 0c   {src=00, dest=ff DATA_NOT_EXPECTING_REPLY}
[NextStationUnknown]
55 ff 01 01 00 00 00 f5   {src=00, dest=01 POLL_FOR_MASTER}
55 ff 01 02 00 00 00 6d   {src=00, dest=02 POLL_FOR_MASTER}
55 ff 01 03 00 00 00 e4   {src=00, dest=03 POLL_FOR_MASTER}
55 ff 01 09 00 00 00 b1   {src=00, dest=09 POLL_FOR_MASTER}
55 ff 01 0a 00 00 00 29   {src=00, dest=0a POLL_FOR_MASTER}
55_ff_02_00_0a_00_00_0a_  {src=0a, dest=00 REPLY_TO_POLL_FOR_MASTER}
[ReceivedReplyToPFM]
55 ff 00 0a 00 00 00 af   {src=00, dest=0a TOKEN}
55_ff_05_00_0a_00_16_6e_01_[Data]
0c_00_0b_06_c0_a8_01_7e_ba_c0_00_05_c0_0c_0c_02_05_f3_6f_19_0b_d0_df_  {src=0a, dest=00 DATA_EXPECTING_REPLY}
[ReceivedDataNeedingReply]
55 ff 07 0a 00 00 00 38   {src=00, dest=0a REPLY_POSTPONED}
55_ff_00_00_0a_00_00_05_  {src=0a, dest=00 TOKEN}
[SendCustomFrame]
55 ff 06 0a 00 00 1b b7 01 20 00 0b 06 c0 a8 01 7e ba c0 ff 30 c0 0c 0c 02 05 f3 6f 19 0b 3e 22 0b b8 3f 5a 4c   {src=00, dest=0a DATA_NOT_EXPECTING_REPLY}
55 ff 00 0a 00 00 00 af   {src=00, dest=0a TOKEN}
55_ff_00_00_0a_00_00_05_  {src=0a, dest=00 TOKEN}
55 ff 00 0a 00 00 00 af   {src=00, dest=0a TOKEN}

MSTP Stats:
-----------
              sentBytes:  2458              recvBytes:  2180             eatAnOctet:    17
             sentFrames:   218             recvFrames:   189      sentPollForMaster:    28
               sentAPDU:    22               recvAPDU:    21 sentReplyPollForMaster:     1
           sentLastType:     0           recvLastType:     0              sentToken:   146
                                          recvInvalid:     1               hasToken:     0
MSTP State:
-----------
       This station (TS):    0              EventCount:    5            master_state:    0
       Next Station (NS):   10              FrameCount:    1              recv_state:    2
       Poll Station (PS):    0              TokenCount:   28              SoleMaster:    0
   Previous Station (AS):   10              RetryCount:    0          SilenceTimerMS: 7384

MSTP Events:
------------
               BadHeader:    0                    Data:    1              EatAnOctet:    0
      NextStationUnknown:    1ReceivedDataNeedingReply:    1    ReceivedInvalidFrame:    0
             ReceivedPFM:    1      ReceivedReplyToPFM:    1     ResetMaintenancePFM:    2
          RetrySendToken:    0         SendCustomFrame:    1      SendMaintenancePFM:    1
```

CHANGELOG.md
# Change Log

## Version 0.0.x

### 0.0.4 (2023-Sep-22)

- Updated the CAS BACnet Stack to version 4.1.19
- Updated the CAS BACnet Stack MSTP stack 4

### 0.0.3 (2023-Jun-28)

- Updated the application function definition for CallbackSetPropertyReal to match the CAS BACnet Stack on non 32bit systems
- Updated the CAS BACnet Stack to version 4.1.16

### 0.0.2 (2020-Sep-25)

- Updated CAS BACnet stack to version 3.19.0

### 0.0.1 (2020-Feb-12)

- Inital release.


BACnetServerMSTPExample/Serial.h
// Serial.h
// http://www.codeguru.com/Cpp/I-N/network/serialcommunications/article.php/c2503

#ifndef __SERIAL_H__
#define __SERIAL_H__
#pragma once


#include <stdio.h>
#include <stdarg.h>
#include <windows.h>


#define FC_DTRDSR       0x01
#define FC_RTSCTS       0x02
#define FC_XONXOFF      0x04
#define ASCII_BEL       0x07
#define ASCII_BS        0x08
#define ASCII_LF        0x0A
#define ASCII_CR        0x0D
#define ASCII_XON       0x11
#define ASCII_XOFF      0x13

typedef void *HANDLE;

class CSerial
{

public:
	CSerial();
	~CSerial();

	bool Open(	int		nPort		= 2, 
				int		nBaud		= 9600,
				char	parity		= 'N', 
				unsigned int databits	= 8, 
				unsigned int stopsbits	= 1 );
	bool Close( void );

	int ReadData( void *, int );
	int SendData( const char *, int );
	int ReadDataWaiting( void );

	bool IsOpened( void ){ return( m_bOpened ); }

protected:
	bool WriteCommByte( unsigned char );
	bool WriteMultipleCommByte( unsigned char * ucByte, int iSize  );

	HANDLE m_hIDComDev;
	OVERLAPPED m_OverlappedRead, m_OverlappedWrite;
	bool m_bOpened;

};

#endif


BACnetServerMSTPExample/CHiTimer.h
/*******************************************************************************
                 CHiTimer.h
********************************************************************************

   Written by:   Steven Smethust 
   Email:        funvill@funvill.com 
   
   No warranty of any kind, express or implied, is included with this
   software; use at your own risk
   
   If you find this code useful, credits would be nice!
   Feel free to distribute it, and email suggested changes to me.
   Free for commercial and non-commercial use
   
********************************************************************************

 Versions
 1.00aA  25 Nov 02 SWS  Created
 1.00aB  11 Ded 06 SWS  Added License 

********************************************************************************/
#ifndef __CHiTimer_h__
#define __CHiTimer_h__

#include <stdint.h>

void CHiTimer_Init();
void CHiTimer_Reset();
uint32_t CHiTimer_DiffTimeMicroSeconds();

#endif // __CHiTimer_h__

BACnetServerMSTPExample/Serial.cpp
// Serial.cpp

#include "Serial.h"


CSerial::CSerial()
{

	memset( &m_OverlappedRead, 0, sizeof( OVERLAPPED ) );
 	memset( &m_OverlappedWrite, 0, sizeof( OVERLAPPED ) );
	m_hIDComDev = NULL;
	m_bOpened = FALSE;

}

CSerial::~CSerial()
{

	Close();

}

bool CSerial::Open( int nPort, int nBaud, char parity, unsigned int  databits, unsigned int  stopsbits )
{
	if( m_bOpened ) return( TRUE );

	char szPort[15];
	char szComParams[50];
	DCB dcb;

	sprintf( szPort, "COM%d", nPort );
	m_hIDComDev = CreateFileA( szPort, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL | FILE_FLAG_OVERLAPPED, NULL );
	if( m_hIDComDev == NULL ) return( FALSE );

	memset( &m_OverlappedRead, 0, sizeof( OVERLAPPED ) );
 	memset( &m_OverlappedWrite, 0, sizeof( OVERLAPPED ) );

	COMMTIMEOUTS CommTimeOuts;
	CommTimeOuts.ReadIntervalTimeout = 0xFFFFFFFF;
	CommTimeOuts.ReadTotalTimeoutMultiplier = 0;
	CommTimeOuts.ReadTotalTimeoutConstant = 0;
	CommTimeOuts.WriteTotalTimeoutMultiplier = 0;
	CommTimeOuts.WriteTotalTimeoutConstant = 5000;
	SetCommTimeouts( m_hIDComDev, &CommTimeOuts );

	sprintf( szComParams, "COM%d:%d,%c,%d,%d" , nPort, nBaud, parity, databits, stopsbits );

	m_OverlappedRead.hEvent = CreateEvent( NULL, TRUE, FALSE, NULL );
	m_OverlappedWrite.hEvent = CreateEvent( NULL, TRUE, FALSE, NULL );

	dcb.DCBlength = sizeof( DCB );
	GetCommState( m_hIDComDev, &dcb );
	dcb.BaudRate = nBaud;
	dcb.ByteSize = 8;
	unsigned char ucSet;
	ucSet = (unsigned char) ( ( FC_RTSCTS & FC_DTRDSR ) != 0 );
	ucSet = (unsigned char) ( ( FC_RTSCTS & FC_RTSCTS ) != 0 );
	ucSet = (unsigned char) ( ( FC_RTSCTS & FC_XONXOFF ) != 0 );
	if( !SetCommState( m_hIDComDev, &dcb ) ||
		!SetupComm( m_hIDComDev, 10000, 10000 ) ||
		m_OverlappedRead.hEvent == NULL ||
		m_OverlappedWrite.hEvent == NULL ){
		DWORD dwError = GetLastError();
		if( m_OverlappedRead.hEvent != NULL ) CloseHandle( m_OverlappedRead.hEvent );
		if( m_OverlappedWrite.hEvent != NULL ) CloseHandle( m_OverlappedWrite.hEvent );
		CloseHandle( m_hIDComDev );
		return( FALSE );
		}

	m_bOpened = TRUE;

	return( m_bOpened );

}

bool CSerial::Close( void )
{

	if( !m_bOpened || m_hIDComDev == NULL ) return( TRUE );

	if( m_OverlappedRead.hEvent != NULL ) CloseHandle( m_OverlappedRead.hEvent );
	if( m_OverlappedWrite.hEvent != NULL ) CloseHandle( m_OverlappedWrite.hEvent );
	CloseHandle( m_hIDComDev );
	m_bOpened = FALSE;
	m_hIDComDev = NULL;

	return( TRUE );

}

bool CSerial::WriteCommByte( unsigned char ucByte )
{
	BOOL bWriteStat;
	DWORD dwBytesWritten;

	bWriteStat = WriteFile( m_hIDComDev, (LPSTR) &ucByte, 1, &dwBytesWritten, &m_OverlappedWrite );
	if( !bWriteStat && ( GetLastError() == ERROR_IO_PENDING ) ){
		if( WaitForSingleObject( m_OverlappedWrite.hEvent, 1000 ) ) dwBytesWritten = 0;
		else{
			GetOverlappedResult( m_hIDComDev, &m_OverlappedWrite, &dwBytesWritten, FALSE );
			m_OverlappedWrite.Offset += dwBytesWritten;
			}
		}

	return( TRUE );

}
bool CSerial::WriteMultipleCommByte( unsigned char * ucByte, int iSize  )
{
	if( ucByte == NULL || iSize <= 0 ) 
	{
		return false; 
	}
	BOOL bWriteStat;
	DWORD dwBytesWritten;

	bWriteStat = WriteFile( m_hIDComDev, (LPCVOID) ucByte, (DWORD)iSize, &dwBytesWritten, &m_OverlappedWrite );
	if( !bWriteStat && ( GetLastError() == ERROR_IO_PENDING ) ){
		if( WaitForSingleObject( m_OverlappedWrite.hEvent, 1000 ) ) dwBytesWritten = 0;
		else{
			GetOverlappedResult( m_hIDComDev, &m_OverlappedWrite, &dwBytesWritten, FALSE );
			m_OverlappedWrite.Offset += dwBytesWritten;
			}
		}

	return( TRUE );

}

int CSerial::SendData( const char *buffer, int size )
{
	if( !m_bOpened || m_hIDComDev == NULL ) return( 0 );


	
	if ( WriteMultipleCommByte( (unsigned char *)buffer, size ) ) {
		return size ; 
	} else {
		return 0 ; 
	}
	
	/*

	DWORD dwBytesWritten = 0;
	int i;
	for( i=0; i<size; i++ ){
		WriteCommByte( buffer[i] );
		dwBytesWritten++;
		}

	return( (int) dwBytesWritten );
	*/
	

}

int CSerial::ReadDataWaiting( void )
{

	if( !m_bOpened || m_hIDComDev == NULL ) return( 0 );

	DWORD dwErrorFlags;
	COMSTAT ComStat;

	ClearCommError( m_hIDComDev, &dwErrorFlags, &ComStat );

	return( (int) ComStat.cbInQue );

}

int CSerial::ReadData( void *buffer, int limit )
{
	if( !m_bOpened || m_hIDComDev == NULL ) return( 0 );

	BOOL bReadStatus;
	DWORD dwBytesRead, dwErrorFlags;
	COMSTAT ComStat;

	ClearCommError( m_hIDComDev, &dwErrorFlags, &ComStat );
	if( !ComStat.cbInQue ) return( 0 );

	dwBytesRead = (DWORD) ComStat.cbInQue;
	if( limit < (int) dwBytesRead ) dwBytesRead = (DWORD) limit;

	bReadStatus = ReadFile( m_hIDComDev, buffer, dwBytesRead, &dwBytesRead, &m_OverlappedRead );
	if( !bReadStatus ){
		if( GetLastError() == ERROR_IO_PENDING ){
			WaitForSingleObject( m_OverlappedRead.hEvent, 2000 );
			return( (int) dwBytesRead );
			}
		return( 0 );
		}

	return( (int) dwBytesRead );

}



BACnetServerMSTPExample/CHiTimer.cpp
/*******************************************************************************
                 CHiTimer.cpp
********************************************************************************

   Written by:   Steven Smethust 
   Email:        funvill@funvill.com 
   
   No warranty of any kind, express or implied, is included with this
   software; use at your own risk
   
   If you find this code useful, credits would be nice!
   Feel free to distribute it, and email suggested changes to me.
   Free for commercial and non-commercial use
   
********************************************************************************

 Versions
 1.00aA  25 Nov 02 SWS  Created
 1.00aB  11 Ded 06 SWS  Added License 

********************************************************************************/
 
#include "CHiTimer.h"
#include <chrono> 

std::chrono::high_resolution_clock::time_point CHiTimer_start;
std::chrono::high_resolution_clock::time_point CHiTimer_stop;

void CHiTimer_Init() {
    CHiTimer_Reset(); 
}
void CHiTimer_Reset() {
    CHiTimer_start = std::chrono::high_resolution_clock::now();
}
uint32_t CHiTimer_DiffTimeMicroSeconds() {
    CHiTimer_stop = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::microseconds> (CHiTimer_stop - CHiTimer_start);
    

    // Return a minimum of 10 micro seconds even if the count is zero. 
    uint32_t ret = (uint32_t)microseconds.count();
    return ret < 10 ? 10 : ret;
}

BACnetServerMSTPExample/CASBACnetStackExampleDatabase.h
/*
 * BACnet Server MSTP Example C++
 * ----------------------------------------------------------------------------
 * CASBACnetStackExampleDatabase.h
 * 
 * The CASBACnetStackExampleDatabase is a data storage that contains 
 * some example data used in the server example. This data is represented by BACnet 
 * objects for this server example. There will be one object of each type currently 
 * supported by the CASBACnetStack.
 * 
 * The database will include the following:
 *	- present value
 *	- name
 *	- for outputs priority array (bool and value)
 * 
 * Created by: Steven Smethurst
*/


#ifndef __CASBACnetStackExampleDatabase_h__
#define __CASBACnetStackExampleDatabase_h__

#include <string>
#include <vector>
#include <stdint.h>
#include <string.h>
#include <map>

// Base class for all object types. 
class ExampleDatabaseBaseObject
{
	public:
		// Const
		static const uint8_t PRIORITY_ARRAY_LENGTH = 16 ; 

		// All objects will have the following properties 
		std::string objectName ; 
		uint32_t instance ; 
};

class ExampleDatabaseAnalogValue : public ExampleDatabaseBaseObject 
{
	public:
		float presentValue;		
};
class ExampleDatabaseDevice : public ExampleDatabaseBaseObject 
{
	public:
		int UTCOffset;
		int64_t currentTimeOffset;
		std::string description;
		uint32_t systemStatus;

		// MSTP 
		uint8_t serialPort;
		uint32_t baudRate;
		uint8_t macAddress;
		uint8_t maxInfoFrames;
		uint8_t maxMaster;
};

class ExampleDatabase {

	public:
		ExampleDatabaseAnalogValue analogValue;
		ExampleDatabaseDevice device;

	// Constructor / Deconstructor
	ExampleDatabase();
	~ExampleDatabase();

	// Set all the objects to have a default value. 
	void Setup();

	// Update the values as needed 
	void Loop(); 

	// Helper Functions	
	void LoadNetworkPortProperties();

};


#endif // __CASBACnetStackExampleDatabase_h__


BACnetServerMSTPExample/CASBACnetStackExampleDatabase.cpp
/*
 * BACnet Server MSTP Example C++
 * ----------------------------------------------------------------------------
 * CASBACnetStackExampleDatabase.cpp
 * 
 * Sets up object names and properties in the database.
 * 
 * Created by: Steven Smethurst
*/

#include "CASBACnetStackExampleDatabase.h"

#include <time.h> // time()
#ifdef _WIN32 
#include <winsock2.h>
#include <iphlpapi.h>
#pragma comment(lib, "IPHLPAPI.lib")

#define MALLOC(x) HeapAlloc(GetProcessHeap(), 0, (x))
#define FREE(x) HeapFree(GetProcessHeap(), 0, (x))
#endif // _WIN32 

ExampleDatabase::ExampleDatabase() {
	this->Setup();
}

ExampleDatabase::~ExampleDatabase() {
	this->Setup();
}

void ExampleDatabase::Setup() {

	this->device.instance = 389999;
	this->device.objectName = "Device Rainbow";
	this->device.UTCOffset = 0;
	this->device.currentTimeOffset = 0;
	this->device.description = "Chipkin test BACnet IP Server device";
	// BACnetDeviceStatus ::= ENUMERATED { operational (0), operational-read-only (1), download-required (2), 
	// download-in-progress (3), non-operational (4), backup-in-progress (5) } 
	this->device.systemStatus = 0; // operational (0), non-operational (4)

	// MSTP 
	this->device.baudRate = 9600; 
	this->device.macAddress = 0;
	this->device.maxInfoFrames = 1; // Default=1;
	this->device.maxMaster = 127; // Default=127;

	// Objects
	this->analogValue.instance = 2;
	this->analogValue.objectName = "AnalogValue Diamond";
	this->analogValue.presentValue = 5.43f;

	this->LoadNetworkPortProperties() ; 
}

void ExampleDatabase::LoadNetworkPortProperties() {

	// This function loads the Network port property values needed.
	// It uses system functions to get values like the IP Address and stores them
	// in the example database

	// ToDo: 
}


void ExampleDatabase::Loop() {
}




BACnetServerMSTPExample/CASBACnetStackExampleConstants.h
/*
 * BACnet Server MSTP Example C++
 * ----------------------------------------------------------------------------
 * CASBACnetStackExampleConstants.h
 * 
 * This is a fully static class that contains all the constants used by the example
 * This includes Object Types, Property Identifiers, etc
 * 
 * Created by: Steven Smethurst
*/

#ifndef __CASBACnetStackExampleConstants_h__
#define __CASBACnetStackExampleConstants_h__

/*

*/
class CASBACnetStackExampleConstants {

public:

	// CAS BACnet Stack network type
	static const uint8_t NETWORK_TYPE_IP = 0;
	static const uint8_t NETWORK_TYPE_MSTP = 1;


	// General Constants
	static const uint32_t NETWORK_PORT_LOWEST_PROTOCOL_LAYER = 4194303;

	// Object Types
	static const uint16_t OBJECT_TYPE_ANALOG_INPUT = 0;
	static const uint16_t OBJECT_TYPE_ANALOG_OUTPUT = 1;
	static const uint16_t OBJECT_TYPE_ANALOG_VALUE = 2;
	static const uint16_t OBJECT_TYPE_BINARY_INPUT = 3;
	static const uint16_t OBJECT_TYPE_BINARY_OUTPUT = 4;
	static const uint16_t OBJECT_TYPE_BINARY_VALUE = 5;
	//static const uint16_t OBJECT_TYPE_CALENDAR = 6;
	//static const uint16_t OBJECT_TYPE_COMMAND = 7;
	static const uint16_t OBJECT_TYPE_DEVICE = 8;
	//static const uint16_t OBJECT_TYPE_EVENT_ENROLLMENT = 9;
	//static const uint16_t OBJECT_TYPE_FILE = 10;
	//static const uint16_t OBJECT_TYPE_GROUP = 11;
	//static const uint16_t OBJECT_TYPE_LOOP = 12;
	static const uint16_t OBJECT_TYPE_MULTI_STATE_INPUT = 13;
	static const uint16_t OBJECT_TYPE_MULTI_STATE_OUTPUT = 14;
	//static const uint16_t OBJECT_TYPE_NOTIFICATION_CLASS = 15;
	//static const uint16_t OBJECT_TYPE_PROGRAM = 16;
	//static const uint16_t OBJECT_TYPE_SCHEDULE = 17;
	//static const uint16_t OBJECT_TYPE_AVERAGING = 18;
	static const uint16_t OBJECT_TYPE_MULTI_STATE_VALUE = 19;
	static const uint16_t OBJECT_TYPE_TREND_LOG = 20;
	//static const uint16_t OBJECT_TYPE_LIFE_SAFETY_POINT = 21;
	//static const uint16_t OBJECT_TYPE_LIFE_SAFETY_ZONE = 22;
	//static const uint16_t OBJECT_TYPE_ACCUMULATOR = 23;
	//static const uint16_t OBJECT_TYPE_PULSE_CONVERTER = 24;
	//static const uint16_t OBJECT_TYPE_EVENT_LOG = 25;
	//static const uint16_t OBJECT_TYPE_GLOBAL_GROUP = 26;
	//static const uint16_t OBJECT_TYPE_TREND_LOG_MULTIPLE = 27;
	//static const uint16_t OBJECT_TYPE_LOAD_CONTROL = 28;
	//static const uint16_t OBJECT_TYPE_STRUCTURED_VIEW = 29;
	//static const uint16_t OBJECT_TYPE_ACCESS_DOOR = 30;
	//static const uint16_t OBJECT_TYPE_TIMER = 31;
	//static const uint16_t OBJECT_TYPE_ACCESS_CREDENTIAL = 32;
	//static const uint16_t OBJECT_TYPE_ACCESS_POINT = 33;
	//static const uint16_t OBJECT_TYPE_ACCESS_RIGHTS = 34;
	//static const uint16_t OBJECT_TYPE_ACCESS_USER = 35;
	//static const uint16_t OBJECT_TYPE_ACCESS_ZONE = 36;
	//static const uint16_t OBJECT_TYPE_CREDENTIAL_DATA_INPUT = 37;
	//static const uint16_t OBJECT_TYPE_NETWORK_SECURITY = 38;
	static const uint16_t OBJECT_TYPE_BITSTRING_VALUE = 39;
	static const uint16_t OBJECT_TYPE_CHARACTERSTRING_VALUE = 40;
	//static const uint16_t OBJECT_TYPE_DATEPATTERN_VALUE = 41;
	static const uint16_t OBJECT_TYPE_DATE_VALUE = 42;
	//static const uint16_t OBJECT_TYPE_DATETIMEPATTERN_VALUE = 43;
	//static const uint16_t OBJECT_TYPE_DATETIME_VALUE = 44;
	static const uint16_t OBJECT_TYPE_INTEGER_VALUE = 45;
	static const uint16_t OBJECT_TYPE_LARGE_ANALOG_VALUE = 46;
	static const uint16_t OBJECT_TYPE_OCTETSTRING_VALUE = 47;
	static const uint16_t OBJECT_TYPE_POSITIVE_INTEGER_VALUE = 48;
	//static const uint16_t OBJECT_TYPE_TIMEPATTERN_VALUE = 49;
	static const uint16_t OBJECT_TYPE_TIME_VALUE = 50;
	//static const uint16_t OBJECT_TYPE_NOTIFICATION_FORWARDER = 51;
	//static const uint16_t OBJECT_TYPE_ALERT_ENROLLMENT = 52;
	//static const uint16_t OBJECT_TYPE_CHANNEL = 53;
	//static const uint16_t OBJECT_TYPE_LIGHTING_OUTPUT = 54;
	//static const uint16_t OBJECT_TYPE_BINARY_LIGHTING_OUTPUT = 55;
	static const uint16_t OBJECT_TYPE_NETWORK_PORT = 56;
	//static const uint16_t OBJECT_TYPE_ELEVATOR_GROUP = 57;
	//static const uint16_t OBJECT_TYPE_ESCALATOR = 58;
	//static const uint16_t OBJECT_TYPE_LIFT = 59;

	// Property Identifiers
	static const uint32_t PROPERTY_IDENTIFIER_ALL = 8;
	static const uint32_t PROPERTY_IDENTIFIER_COV_INCURMENT = 22;
	static const uint32_t PROPERTY_IDENTIFIER_DAY_LIGHT_SAVINGS_STATUS = 24;
	static const uint32_t PROPERTY_IDENTIFIER_DESCRIPTION = 28;
	static const uint32_t PROPERTY_IDENTIFIER_LOCAL_DATE = 56;
	static const uint32_t PROPERTY_IDENTIFIER_LOCAL_TIME = 57;
	static const uint32_t PROPERTY_IDENTIFIER_MAX_INFO_FRAMES = 63;
	static const uint32_t PROPERTY_IDENTIFIER_MAX_MASTER = 64;
	static const uint32_t PROPERTY_IDENTIFIER_MAX_PRES_VALUE = 65;
	static const uint32_t PROPERTY_IDENTIFIER_MIN_PRES_VALUE = 69;
	static const uint32_t PROPERTY_IDENTIFIER_NUMBER_OF_STATES = 74;
	static const uint32_t PROPERTY_IDENTIFIER_OBJECT_NAME = 77;	
	static const uint32_t PROPERTY_IDENTIFIER_PRESENT_VALUE = 85;
	static const uint32_t PROPERTY_IDENTIFIER_PRIORITY_ARRAY = 87;
	static const uint32_t PROPERTY_IDENTIFIER_RELIABILITY = 103;
	static const uint32_t PROPERTY_IDENTIFIER_STATE_TEXT = 110;
	static const uint32_t PROPERTY_IDENTIFIER_STATUS_FLAGS = 111;
	static const uint32_t PROPERTY_IDENTIFIER_SYSTEM_STATUS = 112;
	static const uint32_t PROPERTY_IDENTIFIER_UTC_OFFSET = 119;
	static const uint32_t PROPERTY_IDENTIFIER_BIT_TEXT = 343;

	// Network Port Property Identifiers
	static const uint32_t PROPERTY_IDENTIFIER_IP_ADDRESS = 400;
	static const uint32_t PROPERTY_IDENTIFIER_IP_DEFAULT_GATEWAY = 401;
	static const uint32_t PROPERTY_IDENTIFIER_IP_DNS_SERVER = 406;
	static const uint32_t PROPERTY_IDENTIFIER_IP_SUBNET_MASK = 411;
	static const uint32_t PROPERTY_IDENTIFIER_BACNET_IP_UDP_PORT = 412;
	static const uint32_t PROPERTY_IDENTIFIER_LINK_SPEED = 420;
	static const uint32_t PROPERTY_IDENTIFIER_MAC_ADDRESS = 423;
	
	// Services Supported
	static const uint8_t SERVICE_SUBSCRIBE_COV = 5;
	static const uint8_t SERVICE_CREATE_OBJECT = 10;
	static const uint8_t SERVICE_DELETE_OBJECT = 11;
	static const uint8_t SERVICE_READ_PROPERTY_MULTIPLE = 14;
	static const uint8_t SERVICE_WRITE_PROPERTY = 15;
	static const uint8_t SERVICE_WRITE_PROPERTY_MULTIPLE = 16;
	static const uint8_t SERVICE_I_AM = 26;
	static const uint8_t SERVICE_I_HAVE = 27;
	static const uint8_t SERVICE_TIME_SYNCHRONIZATION = 32;
	static const uint8_t SERVICE_READ_RANGE = 35;
	static const uint8_t SERVICE_UTC_TIME_SYNCHRONIZATION = 36;
	static const uint8_t SERVICE_SUBSCRIBE_COV_PROPERTY = 38;
	

	// Error Codes
	static const uint8_t ERROR_NO_SPACE_TO_WRITE_PROPERTY = 20;
	static const uint8_t ERROR_VALUE_OUT_OF_RANGE = 37;

	// Network Type
	static const uint8_t NETWORK_TYPE_IPV4 = 5;

	// Protocol Level
	static const uint8_t PROTOCOL_LEVEL_BACNET_APPLICATION = 2;

	// Priority Array
	static const uint8_t MAX_BACNET_PRIORITY = 16;

	// Trend Log Buffer Size
	static const uint32_t MAX_TREND_LOG_MAX_BUFFER_SIZE = 100;

	// Data types 
	static const uint32_t DATA_TYPE_NULL = 0;
	static const uint32_t DATA_TYPE_BOOLEAN = 1;
	static const uint32_t DATA_TYPE_UNSIGNED_INTEGER = 2;
	static const uint32_t DATA_TYPE_SIGNED_INTEGER = 3;
	static const uint32_t DATA_TYPE_REAL = 4;
	static const uint32_t DATA_TYPE_DOUBLE = 5;
	static const uint32_t DATA_TYPE_OCTET_STRING = 6;
	static const uint32_t DATA_TYPE_CHARACTER_STRING = 7;
	static const uint32_t DATA_TYPE_BIT_STRING = 8;
	static const uint32_t DATA_TYPE_ENUMERATED = 9;
	static const uint32_t DATA_TYPE_DATE = 10;
	static const uint32_t DATA_TYPE_TIME = 11;
	static const uint32_t DATA_TYPE_BACNET_OBJECT_IDENTIFIER = 12;
};

#endif // __CASBACnetStackExampleConstants_h__
